src/races/races.service.ts

```diff
@@ -294,4 +296,534 @@ export class RacesService {
+      }
+      Example Input: "US Senate - CA"
+      Example Output: {
+      }
+      Example Input: "Maine Village Board Chair - Wisconsin"
+      Example Output:
+      {
+           "village": "Maine Village",
+           "county": "Marathon County"
+      }
+      `
+
+      const tool: any = {
+        type: 'function',
+        function: {
+          name: 'extractLocation',
+          description: 'Extract the location from the office name.',
+          parameters: {
+            type: 'object',
+            properties: {},
+          },
+        },
+      }
+      let systemPrompt
+      if (level === 'county') {
+        systemPrompt = countyPrompt
+      } else if (level === 'city') {
+        systemPrompt = cityPrompt
+        tool.function.parameters.properties.city = {
+          type: 'string',
+          description: 'The city name.',
+        }
+      } else if (level === 'town') {
+        systemPrompt = townPrompt
+        tool.function.parameters.properties.town = {
+          type: 'string',
+          description: 'The town name.',
+        }
+      } else if (level === 'township') {
+        systemPrompt = townshipPrompt
+        tool.function.parameters.properties.township = {
+          type: 'string',
+          description: 'The township name.',
+        }
+      } else if (level === 'village') {
+        systemPrompt = villagePrompt
+        tool.function.parameters.properties.village = {
+          type: 'string',
+          description: 'The village name.',
+        }
+      } else {
+        return false
+      }
+
+      // we always try to get the county name
+      tool.function.parameters.properties.county = {
+        type: 'string',
+        description: 'The county name.',
+      }
+
+      const messages = [
+        {
+          role: 'system',
+          content: systemPrompt,
+        },
+        {
+          role: 'user',
+          content: `Input: "${office}"
+          Output:`,
+        },
+      ]
+
+      const toolChoice: any = {
+        type: 'function',
+        function: { name: 'extractLocation' },
+      }
+
+      let completion
+      // todo: once the ai service is up, we can use this code
+      // const completion = await getChatToolCompletion(
+      //   messages,
+      //   0.1,
+      //   0.1,
+      //   tool,
+      //   toolChoice,
+      // )
+
+      this.logger.log(
+        `messages: ${messages}. tool: ${tool}. toolChoice: ${toolChoice}`,
+      )
+
+      // console.log('completion', completion);
+      const content = completion?.content
+      let decodedContent: any = {}
+      try {
+        decodedContent = JSON.parse(content)
+        decodedContent.level = level
+      } catch (e) {
+        console.log('error at extract-location-ai helper', e)
+      }
+      console.log('extract ai location response', decodedContent)
+      return decodedContent
+    } catch (e) {
+      console.log('error at extract-location-ai helper', e)
+      return false
+    }
```

Comment:
Can we please try to target our `try-catch`es?

There's a lot of just native JS functionality here, no service calls, unless I'm missing something... 🤔 

Everything here looks like it can just be handled by either existential checks, or, in the case of the `JSON.parse()`, `try-catch`ing _just_ around that one operation to protect against a non-JSON string somehow getting through.